File: TiffinBox/app/src/main/java/com/tiff/tiffinbox/authentication/viewModel/AuthViewModel.java
package com.tiff.tiffinbox.authentication.viewModel;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.view.View;
import android.widget.ProgressBar;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.Query;
import com.google.firebase.database.ValueEventListener;
import com.tiff.tiffinbox.Customer.ui.Customer;
import com.tiff.tiffinbox.Seller.AddView;
import com.tiff.tiffinbox.authentication.model.User;

import es.dmoral.toasty.Toasty;

public class AuthViewModel extends ViewModel implements ValueEventListener {

    public MutableLiveData<User> user = new MutableLiveData<>();
    public MutableLiveData<User> userMutableLiveData;
    private Context context;
    private String data, data2, email;
    public static String UT;
    SharedPreferences.Editor editor;
    SharedPreferences sharedPref;
    private FirebaseAuth firebaseAuth;
    private FirebaseUser firebaseUser;
    private DatabaseReference mDatabase;
    private Query querySeller, queryCustomer;
    private DatabaseReference mFirebasedataRefSell;
    private DatabaseReference mFirebasedataRefCust;
    private DatabaseReference dfOnline;
    private DatabaseReference dfOnline2;

    public MutableLiveData<User> getUser(Context context) {
        if (userMutableLiveData == null) {
            userMutableLiveData = new MutableLiveData<>();
            sharedPref = context.getSharedPreferences("UserType", Context.MODE_PRIVATE);
            editor =  sharedPref.edit();
            firebaseAuth = FirebaseAuth.getInstance();
            mDatabase = FirebaseDatabase.getInstance().getReference();
            mFirebasedataRefSell  = FirebaseDatabase.getInstance().getReference();
            mFirebasedataRefCust  = FirebaseDatabase.getInstance().getReference();
            this.context = context;
        }
        return userMutableLiveData;
    }

    public void submitEmail(String email){
        firebaseAuth.sendPasswordResetEmail(email).addOnCompleteListener(new OnCompleteListener<Void>() {
            @Override
            public void onComplete(@NonNull Task<Void> task) {
                if (task.isSuccessful()) {
                    User loginUser = new User(email);
                    userMutableLiveData.postValue(loginUser);
                    Toasty.success(context, "Password Link sent to your E-Mail, Please check your E-Mail", Toast.LENGTH_LONG).show();
                } else {
                    User loginUser = new User(email);
                    userMutableLiveData.postValue(loginUser);
                    Toasty.error(context, task.getException().getMessage(), Toast.LENGTH_LONG).show();
                }
            }
        });
    }

    public void register(ProgressBar progressBar, Activity activity, String name, String mobile, String email, String password, String address, String userType, String imageURL,
                         String status, String username, String search){
        firebaseAuth.createUserWithEmailAndPassword(email, password).addOnCompleteListener(activity, new OnCompleteListener<AuthResult>() {
            @Override
            public void onComplete(@NonNull Task<AuthResult> task) {
                progressBar.setVisibility(View.GONE);
                if (task.isSuccessful()) {
                    firebaseAuth.getCurrentUser().sendEmailVerification().addOnCompleteListener(new OnCompleteListener<Void>() {
                        @Override
                        public void onComplete(@NonNull Task<Void> task) {
                            if (task.isSuccessful()) {
                                FirebaseUser firebaseUser = firebaseAuth.getCurrentUser();
                                String userUid = firebaseUser.getUid();
                                User registerUser = new User(name, mobile, email, password, address, userType, imageURL, status, userUid, username, search);
                                userMutableLiveData.postValue(registerUser);
                                String uid = firebaseAuth.getCurrentUser().getUid();
                                mDatabase.child(userType).child(uid).setValue(registerUser);
                                Toasty.success(context, "Registered Successfully Please check your E-Mail for verification", Toast.LENGTH_LONG).show();
                            } else {
                                Toasty.error(context, task.getException().getMessage(), Toast.LENGTH_LONG).show();
                            }
                        }
                    });

                } else {
                    Toasty.error(context, "SignUp Unsuccessful " + task.getException().getMessage(), Toast.LENGTH_LONG).show();
                }
            }
        });
    }

    public void login(ProgressBar progressBar, Activity activity, String email, String password){
        this.email = email;
        progressBar.setVisibility(View.VISIBLE);
        firebaseAuth.signInWithEmailAndPassword(email, password).addOnCompleteListener(activity, new OnCompleteListener<AuthResult>() {
            @Override
            public void onComplete(@NonNull Task<AuthResult> task) {
                progressBar.setVisibility(View.GONE);
                if (task.isSuccessful()) {
                    if (firebaseAuth.getCurrentUser().isEmailVerified()){
                        querySeller = mFirebasedataRefSell.child("Seller").orderByChild("email").equalTo(email);
                        queryCustomer = mFirebasedataRefCust.child("Customer").orderByChild("email").equalTo(email);
                        triggerQuey();
                        Toasty.success(context, "Login Success", Toast.LENGTH_SHORT).show();
                    }else {
                        Toasty.error(context, "Please verify your E-Mail address", Toast.LENGTH_LONG).show();
                    }
                } else {
                    Toasty.error(context, "SignIn Failed " + task.getException().getMessage(), Toast.LENGTH_LONG).show();
                }
            }
        });
    }

    public void authCheck(){
        firebaseUser = firebaseAuth.getCurrentUser();
        if (firebaseUser!=null&&firebaseAuth.getCurrentUser().isEmailVerified()){
            dfOnline = FirebaseDatabase.getInstance().getReference();
            dfOnline2 = FirebaseDatabase.getInstance().getReference();

            Query myTopPostsQuery = dfOnline.child("Customer").orderByChild("email").equalTo(firebaseUser.getEmail());
            Query myTopPostsQuery2 = dfOnline2.child("Seller").orderByChild("email").equalTo(firebaseUser.getEmail());

            if (sharedPref.getString("UT",null).equals("Customer")){
                myTopPostsQuery.addValueEventListener(new ValueEventListener() {
                    @Override
                    public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                        if (dataSnapshot.getKey().equals("Customer")){
                            dfOnline.child("Customer").orderByChild("email").equalTo(firebaseUser.getEmail()).removeEventListener(this);
                            context.startActivity(new Intent(context, Customer.class));
                            ((Activity)context).finish();
                        }
                    }
                    @Override
                    public void onCancelled(@NonNull DatabaseError databaseError) {

                    }
                });
            }else if (sharedPref.getString("UT",null).equals("Seller")) {
                myTopPostsQuery2.addValueEventListener(new ValueEventListener() {
                    @Override
                    public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                        if (dataSnapshot.getKey().equals("Seller")) {
                            dfOnline2.child("Seller").orderByChild("email").equalTo(firebaseUser.getEmail()).removeEventListener(this);
                            context.startActivity(new Intent(context, AddView.class));
                            ((Activity)context).finish();
                        }
                    }
                    @Override
                    public void onCancelled(@NonNull DatabaseError databaseError) {

                    }
                });
            }
        }
    }

    private void triggerQuey() {
        if (queryCustomer != null && querySeller != null) {
            querySeller.addValueEventListener(this);
            queryCustomer.addValueEventListener(this);
        }
    }

    @Override
    public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
        if (dataSnapshot.getValue() != null) {
            String key = dataSnapshot.getKey();
            if (key.equals("Customer")) {
                data = dataSnapshot.getValue().toString();
                editor.putString("UT", "Customer");
                editor.commit();
                mFirebasedataRefCust.child("Customer").orderByChild("email").equalTo(email).removeEventListener(this);
                context.startActivity(new Intent(context, Customer.class));
                UT = "Customer";
                ((Activity)context).finish();
            } else if (key.equals("Seller")){
                UT = "Seller";
                editor.putString("UT", "Seller");
                editor.commit();
                data2 = dataSnapshot.getValue().toString();
                mFirebasedataRefSell.child("Seller").orderByChild("email").equalTo(email).removeEventListener(this);
                context.startActivity(new Intent(context, AddView.class));
                ((Activity)context).finish();
            }
        }
    }
    @Override
    public void onCancelled(@NonNull DatabaseError databaseError) {
        Toast.makeText(context,"failed "+data2,Toast.LENGTH_SHORT).show();
    }
}

File: TiffinBox/app/src/main/java/com/tiff/tiffinbox/authentication/viewModel/AuthViewModel.kt
package com.tiff.tiffinbox.authentication.viewModel

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.view.View
import android.widget.ProgressBar
import android.widget.Toast
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.database.*
import com.tiff.tiffinbox.Customer.ui.Customer
import com.tiff.tiffinbox.Seller.AddView
import com.tiff.tiffinbox.authentication.model.User
import es.dmoral.toasty.Toasty

class AuthViewModel : ViewModel(), ValueEventListener {
    @JvmField
    var user = MutableLiveData<User>()
    var userMutableLiveData: MutableLiveData<User>? = null
    private var context: Context? = null
    private var data: String? = null
    private var data2: String? = null
    private var email: String? = null
    var editor: SharedPreferences.Editor? = null
    var sharedPref: SharedPreferences? = null
    private var firebaseAuth: FirebaseAuth? = null
    private var firebaseUser: FirebaseUser? = null
    private var mDatabase: DatabaseReference? = null
    private var querySeller: Query? = null
    private var queryCustomer: Query? = null
    private var mFirebasedataRefSell: DatabaseReference? = null
    private var mFirebasedataRefCust: DatabaseReference? = null
    private var dfOnline: DatabaseReference? = null
    private var dfOnline2: DatabaseReference? = null
    fun getUser(context: Context): MutableLiveData<User> {
        if (userMutableLiveData == null) {
            userMutableLiveData = MutableLiveData()
            sharedPref = context.getSharedPreferences("UserType", Context.MODE_PRIVATE)
            editor = sharedPref!!.edit()
            firebaseAuth = FirebaseAuth.getInstance()
            mDatabase = FirebaseDatabase.getInstance().reference
            mFirebasedataRefSell = FirebaseDatabase.getInstance().reference
            mFirebasedataRefCust = FirebaseDatabase.getInstance().reference
            this.context = context
        }
        return userMutableLiveData!!
    }

    fun submitEmail(email: String?) {
        firebaseAuth!!.sendPasswordResetEmail(email!!).addOnCompleteListener { task ->
            if (task.isSuccessful) {
                val loginUser = User(email)
                userMutableLiveData!!.postValue(loginUser)
                Toasty.success(context!!, "Password Link sent to your E-Mail, Please check your E-Mail", Toast.LENGTH_LONG).show()
            } else {
                val loginUser = User(email)
                userMutableLiveData!!.postValue(loginUser)
                Toasty.error(context!!, task.exception!!.message!!, Toast.LENGTH_LONG).show()
            }
        }
    }

    fun register(progressBar: ProgressBar, activity: Activity?, name: String?, mobile: String?, email: String?, password: String?, address: String?, userType: String?, imageURL: String?,
                 status: String?, username: String?, search: String?) {
        firebaseAuth!!.createUserWithEmailAndPassword(email!!, password!!).addOnCompleteListener(activity!!) { task ->
            progressBar.visibility = View.GONE
            if (task.isSuccessful) {
                firebaseAuth!!.currentUser!!.sendEmailVerification().addOnCompleteListener { task ->
                    if (task.isSuccessful) {
                        val firebaseUser = firebaseAuth!!.currentUser
                        val userUid = firebaseUser!!.uid
                        val registerUser = User(name, mobile, email, password, address, userType, imageURL, status, userUid, username, search)
                        userMutableLiveData!!.postValue(registerUser)
                        val uid = firebaseAuth!!.currentUser!!.uid
                        mDatabase!!.child(userType!!).child(uid).setValue(registerUser)
                        Toasty.success(context!!, "Registered Successfully Please check your E-Mail for verification", Toast.LENGTH_LONG).show()
                    } else {
                        Toasty.error(context!!, task.exception!!.message!!, Toast.LENGTH_LONG).show()
                    }
                }
            } else {
                Toasty.error(context!!, "SignUp Unsuccessful " + task.exception!!.message, Toast.LENGTH_LONG).show()
            }
        }
    }

    fun login(progressBar: ProgressBar, activity: Activity?, email: String?, password: String?) {
        this.email = email
        progressBar.visibility = View.VISIBLE
        firebaseAuth!!.signInWithEmailAndPassword(email!!, password!!).addOnCompleteListener(activity!!) { task ->
            progressBar.visibility = View.GONE
            if (task.isSuccessful) {
                if (firebaseAuth!!.currentUser!!.isEmailVerified) {
                    querySeller = mFirebasedataRefSell!!.child("Seller").orderByChild("email").equalTo(email)
                    queryCustomer = mFirebasedataRefCust!!.child("Customer").orderByChild("email").equalTo(email)
                    triggerQuey()
                    Toasty.success(context!!, "Login Success", Toast.LENGTH_SHORT).show()
                } else {
                    Toasty.error(context!!, "Please verify your E-Mail address", Toast.LENGTH_LONG).show()
                }
            } else {
                Toasty.error(context!!, "SignIn Failed " + task.exception!!.message, Toast.LENGTH_LONG).show()
            }
        }
    }

    fun authCheck() {
        firebaseUser = firebaseAuth!!.currentUser
        if (firebaseUser != null && firebaseAuth!!.currentUser!!.isEmailVerified) {
            dfOnline = FirebaseDatabase.getInstance().reference
            dfOnline2 = FirebaseDatabase.getInstance().reference
            val myTopPostsQuery = dfOnline!!.child("Customer").orderByChild("email").equalTo(firebaseUser!!.email)
            val myTopPostsQuery2 = dfOnline2!!.child("Seller").orderByChild("email").equalTo(firebaseUser!!.email)
            if (sharedPref!!.getString("UT", null) == "Customer") {
                myTopPostsQuery.addValueEventListener(object : ValueEventListener {
                    override fun onDataChange(dataSnapshot: DataSnapshot) {
                        if (dataSnapshot.key == "Customer") {
                            dfOnline!!.child("Customer").orderByChild("email").equalTo(firebaseUser!!.email).removeEventListener(this)
                            context!!.startActivity(Intent(context, Customer::class.java))
                            (context as Activity?)!!.finish()
                        }
                    }

                    override fun onCancelled(databaseError: DatabaseError) {}
                })
            } else if (sharedPref!!.getString("UT", null) == "Seller") {
                myTopPostsQuery2.addValueEventListener(object : ValueEventListener {
                    override fun onDataChange(dataSnapshot: DataSnapshot) {
                        if (dataSnapshot.key == "Seller") {
                            dfOnline2!!.child("Seller").orderByChild("email").equalTo(firebaseUser!!.email).removeEventListener(this)
                            context!!.startActivity(Intent(context, AddView::class.java))
                            (context as Activity?)!!.finish()
                        }
                    }

                    override fun onCancelled(databaseError: DatabaseError) {}
                })
            }
        }
    }

    private fun triggerQuey() {
        if (queryCustomer != null && querySeller != null) {
            querySeller!!.addValueEventListener(this)
            queryCustomer!!.addValueEventListener(this)
        }
    }

    override fun onDataChange(dataSnapshot: DataSnapshot) {
        if (dataSnapshot.value != null) {
            val key = dataSnapshot.key
            if (key == "Customer") {
                data = dataSnapshot.value.toString()
                editor!!.putString("UT", "Customer")
                editor!!.commit()
                mFirebasedataRefCust!!.child("Customer").orderByChild("email").equalTo(email).removeEventListener(this)
                context!!.startActivity(Intent(context, Customer::class.java))
                UT = "Customer"
                (context as Activity?)!!.finish()
            } else if (key == "Seller") {
                UT = "Seller"
                editor!!.putString("UT", "Seller")
                editor!!.commit()
                data2 = dataSnapshot.value.toString()
                mFirebasedataRefSell!!.child("Seller").orderByChild("email").equalTo(email).removeEventListener(this)
                context!!.startActivity(Intent(context, AddView::class.java))
                (context as Activity?)!!.finish()
            }
        }
    }

    override fun onCancelled(databaseError: DatabaseError) {
        Toast.makeText(context, "failed $data2", Toast.LENGTH_SHORT).show()
    }

    companion object {
        var UT: String? = null
    }
}